client/client-http/src/main/java/software/amazon/smithy/java/client/http/compression/GzipCompressingInputStream.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.java.client.http.compression;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.zip.GZIPOutputStream;
+
+final class GzipCompressingInputStream extends InputStream {
+
+    private final InputStream source;
+    private byte[] compressedData;
+    private int position = 0;
+
+    public GzipCompressingInputStream(InputStream source) {
+        this.source = source;
+    }
+
+    @Override
+    public int read() throws IOException {
+        ensureCompressed();
+        return position < compressedData.length ? compressedData[position++] & 0xFF : -1;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        ensureCompressed();
+        var available = compressedData.length - position;
+        if (available <= 0)
+            return -1;
+
+        var toRead = Math.min(len, available);
+        System.arraycopy(compressedData, position, b, off, toRead);
+        position += toRead;
+        return toRead;
+    }
+
+    private void ensureCompressed() throws IOException {
+        if (compressedData == null) {
+            var buffer = new ByteArrayOutputStream();
+            try (GZIPOutputStream gzip = new GZIPOutputStream(buffer)) {
+                source.transferTo(gzip);
+            }
```

Comment:
This reads all the contents from source and transfers them to the gzip output stream.